zcash_client_sqlite/src/lib.rs

```diff
@@ -242,13 +247,63 @@ impl fmt::Display for ReceivedNoteId {
 pub struct UtxoId(pub i64);
 
 /// A newtype wrapper for sqlite primary key values for the transactions table.
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
 struct TxRef(pub i64);
 
+/// A newtype wrapper for sqlite primary key values for the addresses table.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+struct AddressRef(pub(crate) i64);
+
+/// A data structure that can be used to configure custom gap limits for use in transparent address
+/// rotation.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg(feature = "transparent-inputs")]
+pub struct GapLimits {
+    external: u32,
+    transparent_internal: u32,
+    ephemeral: u32,
```

Comment:
As previously commented, I think that compatibility requires we keep the ephemeral gap limit at 20. I'm also concerned about the risk of inadvertently creating similar compatibility problems for external and transparent_internal addresses, since it is incorrect to reduce any of these limits for a given wallet instance.

Reply:
To clarify the compatibility argument: existing wallets might have gaps in their ephemeral address indices of up to 19. Sure, it's unlikely, but those wallets could have continued to allocate addresses past a gap of that size.

Reply:
I still think that this is fine, because it's even lower-probability that funds were ever _returned_ to these addresses - in order for a user to encounter a problem, they'd both have to have multiple failures _and_ then that address would have to be used for a return of funds. If after this change, a user contacts us about getting access to returned funds, _then_ I think we should consider bumping up the gap limit.

Reply:
The other thing that is relevant here is that we're making some assumptions about the generation of _all_ transparent addresses in other wallets. We only support a gap limit of 10 addresses, because we need extra space below the industry-standard of 20 addresses for handling the case where, in the future, we enable UA rotation in a way that more strongly protects against linkability. So it's (remotely) possible that some users might not find all of their funds with the current gap limit approach; I think we should just take that as a known limitation - if users really need to, they can manually fill in the gap to get to where all their addresses are visible.

Reply:
> in order for a user to encounter a problem, they'd both have to have multiple failures and then that address would have to be used for a return of funds

I don't think that's true? To have a problem, they'd have to have had 3 or more consecutive failures to send at some point in the lifetime of the old wallet (which could just be a temporary loss of connectivity), followed by a return of funds on *any subsequent address*, and then they need to restore. They restore and they don't get the returned funds, because scanning will stop before that point.

Reply:
I think that we should have a consistent approach to "the wallet is missing funds due to the gap limit" which is: document the potential problem, and provide the user with an avenue to recover: in this case, that avenue is to send funds to the maximum-allowed ephemeral address to manually advance the gap.

Reply:
I'm sufficiently happy with the compromise of bumping the ephemeral gap limit to 5.